### PR TITLE
#58 Employee Constructor + Pwd Encryption (new PR)

### PR DESCRIPTION
**Pull request by bkuzmic**
For #58:

- Creating constructor for Employee model which sets defaults fields and encrypts password
- Update tests to reflect changes

### CODE REVIEW

**Comment by emilianodellacasa** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
Job [#138](https://github.com/timelesslounge/timelessis/issues/138) is now in scope, role is `REV`

File: timeless/employees/models.py [hunk 1]
```diff
@@ -37,5 +32,20 @@ class Employee(TimestampsMixin, DB.Model):
 
     company = DB.relationship("Company", back_populates="employees")
 
+    def __init__(self, **kwargs):
+        super(Employee, self).__init__(**kwargs)
+        if "password" in kwargs:
+            self.password = bcrypt_sha256.hash(kwargs.get("password"))
+        if "pin_code" in kwargs:
```
Comment (emilianodellacasa) [inline timeless/employees/models.py]: @bkuzmic What happens if pin code is not provided?

**Comment by emilianodellacasa** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
This pull request [#138](https://github.com/timelesslounge/timelessis/issues/138) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/c95b5add-9412-4ba2-bc9b-e45f626b2572); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by emilianodellacasa** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @vladarefiev

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @amihaiemil

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil  ([here](https://github.com/timelesslounge/timelessis/issues/138#issuecomment-461053043))

@emilianodellacasa Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @amihaiemil

**Comment by 0crat** [comment]
This pull request [#138](https://github.com/timelesslounge/timelessis/issues/138) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/3991820e-6f88-43ad-b78b-346c10c574a8); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @vladarefiev

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @amihaiemil

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#138` already assigned to @vladarefiev, can't assign to @vladarefiev

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil ([here](https://github.com/timelesslounge/timelessis/issues/138#issuecomment-461088333))

@emilianodellacasa This pull request [#138](https://github.com/timelesslounge/timelessis/issues/138) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/6a9f3177-d7e4-4a77-bf13-d1985bbcea8b); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil ([here](https://github.com/timelesslounge/timelessis/issues/138#issuecomment-461092141))

@emilianodellacasa Job `gh:timelesslounge/timelessis#138` already assigned to @amihaiemil, can't assign to @amihaiemil

File: timeless/employees/models.py (hunk 1 above)
Comment (bkuzmic) [reply timeless/employees/models.py]: @emilianodellacasa Should I throw KeyError? Also, should I do the same for other non optional parameters, like username? The question is more: do we want to have this kind of validation on the model side or in the controller?

**Comment by bkuzmic** [comment]
@emilianodellacasa I've added a validator decorator to check all mandatory parameters. Do you think this is ok?

File: timeless/employees/models.py (hunk 1 above)
Comment (emilianodellacasa) [reply timeless/employees/models.py]: @bkuzmic Supposedly, the pin code should be assigned randomly. I am ok with having an optional parameter, but the main request was to assign a random value to it;

**Comment by emilianodellacasa** [comment]
@bkuzmic This is a great idea, please add a puzzle to implement the decorator on all models and see my comment;

File: timeless/employees/models.py (hunk 1 above)
Comment (bkuzmic) [reply timeless/employees/models.py]: @emilianodellacasa Ok, I'll generate a random pin number in the constructor.

**Comment by bkuzmic** [comment]
@emilianodellacasa Please take a look if now is ok.

**Comment by emilianodellacasa** [comment]
@bkuzmic Awesome, thanks!

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 21 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#138](https://github.com/timelesslounge/timelessis/issues/138) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review in old PR #122. 
Please confirm that you'll try to find at least three major problems while future reviews

**Comment by amihaiemil** [comment]
@ypshenychka I confirm :)

**Comment by ypshenychka** [comment]
@amihaiemil thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**